### PR TITLE
Add a __using__/1 macro to the Connection module

### PR DESCRIPTION
**Pull request by whatyouhide**
Hey there @fishcakez, I'm not sure we're supposed to start contributing to this project (given its newborn nature) but nobody can get hurt trying, right? :)

Similarly to how `GenServer` in Elixir provides a `__using__/1` macro that defines default implementation for the `:gen_server` behaviour callbacks, I added a `__using__/1` macro to `Connection` that does exactly the same.

The implementation for the `:gen_server` behaviour functions has been taken verbatim from Elixir's `GenServer`. I only added the default implementation for `connect/2` and `disconnect/2`.

Long story short, with this PR you're able to do:

```elixir
defmodule TCPConn do
  use Connection

  # only define functions I care about
end
```

and not get warnings for not having implemented, say, `terminate/2` or `code_change/3`. Let me know what you think :)

### CODE REVIEW

File: lib/connection.ex [hunk 1]
```diff
@@ -291,6 +291,68 @@ defmodule Connection do
...
+    quote location: :keep do
+      @behaviour Connection
+
+      # The default implementations of init/1, handle_call/3, handle_info/2,
+      # handle_cast/2, terminate/2 and code_change/3 have been taken verbatim
+      # from Elixir's GenServer default implementation.
+
+      @doc false
+      def init(args) do
+        {:ok, args}
```
Comment (fishcakez) [inline lib/connection.ex]: `{:ok, args}` is not a valid return for `init/1` in `Connection`. Initially I wanted the behaviour to be a superset of GenServer. However it is important to know whether the process is going to connect (or not) so I decided it should be explicit about what the first step is after `init/1`. The callback returns are:
```elixir
 defcallback init(any) ::
    {:connect, any, any} |
    {:backoff, timeout, any} | {:backoff, timeout, any, timeout | :hibernate} |
    {:noconnect, any} | {:noconnect, any, timeout | :hibernate} |
    :ignore | {:stop, any}
```
What is a good default?

File: lib/connection.ex [hunk 2]
```diff
@@ -291,6 +291,68 @@ defmodule Connection do
...
+
+      @doc false
+      def code_change(_old, state, _extra) do
+        {:ok, state}
+      end
+
+      @doc false
+      def connect(_, _state) do
+        exit(:bad_connect)
+      end
```
Comment (fishcakez) [inline lib/connection.ex]: The first argument is information for that connection attempt. Can you include that in the exit reason, `{:bad_connect, info}`. I am happy not to use the dialyzer trick here because the warning is valid - why use a connection if can't connect?

File: lib/connection.ex [hunk 3]
```diff
@@ -291,6 +291,68 @@ defmodule Connection do
...
+
+      @doc false
+      def disconnect(_, _state) do
+        exit(:bad_disconnect)
+      end
```
Comment (fishcakez) [inline lib/connection.ex]: Same as `connect/2.` Must explicitly disconnect with `{:disconnect, ..}` so its a bug if its called and not implemented.

**Comment by fishcakez** [comment]
Please feel free to contribute. I hope it is feature complete already - except test cases.

**Comment by whatyouhide** [comment]
@fishcakez fixed the things you mentioned and `push --force`'d :) Let me know what you think!

**Comment by fishcakez** [comment]
This is perfect, thanks. I will think about reverting the `:noconnect` returns from `init/1` to `:ok`.
